Reading secrets thru Vault Agent with gzip encoding between Vault agent and Vault server

<!-- Please reserve GitHub issues for bug reports and feature requests.

For questions, the best place to get answers is on our [mailing list](https://groups.google.com/forum/#!forum/vault-tool), as they will get more visibility from experienced users than the issue tracker.

Please note: We take Vault's security and our users' trust very seriously. If you believe you have found a security issue in Vault, please responsibly disclose by contacting us at security@hashicorp.com. Our PGP key is available at [our security page](https://www.hashicorp.com/security/).

-->

**Describe the bug**
Configured the vault agent and when try accessing data thru it, the error is
```
2019-04-18T08:24:27.677Z [INFO]  cache: received request: path=<redacted> method=GET
2019-04-18T08:24:27.677Z [DEBUG] cache.leasecache: forwarding request: path=<redacted> method=GET
2019-04-18T08:24:27.766Z [INFO]  cache.apiproxy: forwarding request: path=<redacted> method=GET
2019-04-18T08:24:27.771Z [ERROR] cache.leasecache: failed to parse response as secret: error="invalid character '\x1f' looking for beginning of value"
```
Executing the query with Curl works, so it's clearly not a problem of permissions

**To Reproduce**
Steps to reproduce the behavior:
1. Run `vault kv put secret/test foo=bar` directly to vault without passing thru the agent
2. Run `vault read secret/data/test` with `export VAULT_AGENT_ADDR=http://127.0.0.1:8100`
3. See error

**Expected behavior**
See the content of the secret

**Environment:**
* Vault Server Version (retrieve with `vault status`):
```
Key                      Value
---                      -----
Recovery Seal Type       shamir
Initialized              true
Sealed                   false
Total Recovery Shares    5
Threshold                3
Version                  1.1.1
Cluster Name             acceptance
Cluster ID               ------
HA Enabled               true
HA Cluster               https://10.244.78.109:8201
HA Mode                  active
```
* Vault CLI Version (retrieve with `vault version`):
`Vault v1.1.1 ('a3dcd63451cf6da1d04928b601bbe9748d53842e')`
* Server Operating System/Architecture:
Kubernetes 1.12.4, CoreOS, Docker 18.06.

**Additional context**
This is due to gzip encoding done by the ingress-nginx in front of Vault Server. If I send the vault agent directly to Vault server this issue is not present

This likely applies to all installation with LB with HTTPS termination in front of Vault

Found the problem with tcpdump. Vault Agent is trying to parse a secret (`command/agent/lease_cache.go` L239) as it was in plain test but the body is coming with gzip encoding. In the Vault cli this case is handled, but it's not in the proxier resp.Body. I'd like to fix this, I just need some pointer

--- a/command/agent/cache/api_proxy.go
+++ b/command/agent/cache/api_proxy.go
@@ -3,11 +3,9 @@ package cache
 import (
 	"context"
 	"fmt"
-	"net/http"
 
 	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/vault/api"
-	"github.com/hashicorp/vault/sdk/helper/strutil"
 )
 
 // APIProxy is an implementation of the proxier interface that is used to
@@ -40,14 +38,10 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, 
 	client.SetToken(req.Token)
 
 	// http.Transport will transparently request gzip and decompress the response, but only if
-	// the client doesn't manually set the header (in which case the client has to handle
-	// the decompression. If gzip has already been set, remove it to avoid triggering the manual
-	// handling requirement.
-	h := clone(req.Request.Header)
-	if v, ok := h["Accept-Encoding"]; ok {
-		h["Accept-Encoding"] = strutil.StrListDelete(v, "gzip")
-	}
-	client.SetHeaders(h)
+	// the client doesn't manually set the header. Removing any Accept-Encoding header allows the
+	// transparent compression to occur.
+	req.Request.Header.Del("Accept-Encoding")
+	client.SetHeaders(req.Request.Header)
 
 	fwReq := client.NewRequest(req.Request.Method, req.Request.URL.Path)
 	fwReq.BodyBytes = req.RequestBody
@@ -75,30 +69,4 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, 
 
 	// Bubble back the api.Response as well for error checking/handling at the handler layer.
 	return sendResponse, err
-}
-
-// clone returns a copy of h or nil if h is nil.
-//
-// TODO: This is a copy of the new (Header) Clone method in Go 1.13. We should remove this once we update.
-func clone(h http.Header) http.Header {
-	if h == nil {
-		return nil
-	}
-
-	// Find total number of values.
-	nv := 0
-	for _, vv := range h {
-		nv += len(vv)
-	}
-	sv := make([]string, nv) // shared backing array for headers' values
-	h2 := make(http.Header, len(h))
-
-	for k, vv := range h {
-		n := copy(sv, vv)
-		h2[k] = sv[:n:n]
-		sv = sv[n:]
-
-	}
-
-	return h2
 }

--- a/command/agent/cache/api_proxy.go
+++ b/command/agent/cache/api_proxy.go
@@ -7,6 +7,7 @@ import (
 
 	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/vault/api"
+	"github.com/hashicorp/vault/sdk/helper/strutil"
 )
 
 // APIProxy is an implementation of the proxier interface that is used to
@@ -43,8 +44,8 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, 
 	// the decompression. If gzip has already been set, remove it to avoid triggering the manual
 	// handling requirement.
 	h := clone(req.Request.Header)
-	if h.Get("Accept-Encoding") == "gzip" {
-		h.Del("Accept-Encoding")
+	if v, ok := h["Accept-Encoding"]; ok {
+		h["Accept-Encoding"] = strutil.StrListDelete(v, "gzip")
 	}
 	client.SetHeaders(h)
 

--- a/command/agent/cache/api_proxy.go
+++ b/command/agent/cache/api_proxy.go
@@ -3,6 +3,7 @@ package cache
 import (
 	"context"
 	"fmt"
+	"net/http"
 
 	hclog "github.com/hashicorp/go-hclog"
 	"github.com/hashicorp/vault/api"
@@ -36,7 +37,16 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, 
 		return nil, err
 	}
 	client.SetToken(req.Token)
-	client.SetHeaders(req.Request.Header)
+
+	// http.Transport will transparently request gzip and decompress the response, but only if
+	// the client doesn't manually set the header (in which case the client has to handle
+	// the decompression. If gzip has already been set, remove it to avoid triggering the manual
+	// handling requirement.
+	h := clone(req.Request.Header)
+	if h.Get("Accept-Encoding") == "gzip" {
+		h.Del("Accept-Encoding")
+	}
+	client.SetHeaders(h)
 
 	fwReq := client.NewRequest(req.Request.Method, req.Request.URL.Path)
 	fwReq.BodyBytes = req.RequestBody
@@ -64,4 +74,30 @@ func (ap *APIProxy) Send(ctx context.Context, req *SendRequest) (*SendResponse, 
 
 	// Bubble back the api.Response as well for error checking/handling at the handler layer.
 	return sendResponse, err
+}
+
+// clone returns a copy of h or nil if h is nil.
+//
+// TODO: This is a copy of the new (Header) Clone method in Go 1.13. We should remove this once we update.
+func clone(h http.Header) http.Header {
+	if h == nil {
+		return nil
+	}
+
+	// Find total number of values.
+	nv := 0
+	for _, vv := range h {
+		nv += len(vv)
+	}
+	sv := make([]string, nv) // shared backing array for headers' values
+	h2 := make(http.Header, len(h))
+
+	for k, vv := range h {
+		n := copy(sv, vv)
+		h2[k] = sv[:n:n]
+		sv = sv[n:]
+
+	}
+
+	return h2
 }

